### PR TITLE
Replace use of str(pathlib.Path(...)) in codebase

### PR DESCRIPTION
**Pull request by zili55**
I removed `str()` functions from `str(pathlib.Path(...))` or replaced with `os.fsdecode`. `str()` for Pathlib objects was introduced for python 3.5 compatibility.

Refs #592 

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This looks really good - nice work!

I've flagged a few things for follow up; they almost all fall into one of three camps:

1. Calls to stdlib functions where I'm almost certain explicit `fsdecode` conversion isn't required - this is calls like `shutil.rmtree()`
2. Calls to stdlib functions where I *suspect* explicit `fsdecode` conversion isn't required, but I could be wrong - this is mostly in calls to `subprocess` et al, where my gut tells me that Path objects *should* be OK for `cwd`, and *might* be OK for arguments... but I also wouldn't be surprised if they're not. Investigate; if I'm wrong, let me know and move on.
3. Calls where *we* are in control of the external API surface. Wherever possible, the API exposed by Briefcase should be the same as stdlib - "string or path like object". We should be managing any `fsdecode` conversion so that the end-user doesn't need to worry about that detail (to the extent possible).

To be clear - what you've done is great, and works fine (as evidenced by the tests passing). The issues I've flagged in group 1 and 2 are subtle edge cases where the conversions you've done aren't *wrong* - they're just a little over-enthusiastic. 

The problems in group 3 are mostly problems you've inherited, and the conversions you've done are doing exactly the right thing, but since we're doubling down on making our usage of Path correct, we should take the extra time and clean up those uses, too.

File: src/briefcase/commands/create.py [hunk 1]
```diff
@@ -370,8 +371,8 @@ def install_app_support_package(self, app: BaseConfig):
             support_path = self.support_path(app)
             support_path.mkdir(parents=True, exist_ok=True)
             self.shutil.unpack_archive(
-                str(support_filename),
-                extract_dir=str(support_path)
+                os.fsdecode(support_filename),
+                extract_dir=os.fsdecode(support_path)
```
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/commands/create.py:375]: according to docs it should be `os.fsdecode` or `str()`: `Changed in version 3.7: Accepts a path-like object for filename and extract_dir.` [link](https://docs.python.org/3/library/shutil.html#shutil.unpack_archive)

File: src/briefcase/platforms/android/gradle.py [hunk 2]
```diff
@@ -110,11 +111,11 @@ def build_app(self, app: BaseConfig, **kwargs):
                 # Windows needs the full path to `gradlew`; macOS & Linux can find it
                 # via `./gradlew`. For simplicity of implementation, we always provide
                 # the full path.
-                [str(self.gradlew_path(app)), "assembleDebug"],
+                [os.fsdecode(self.gradlew_path(app)), "assembleDebug"],
                 env=self.android_sdk.env,
                 # Set working directory so gradle can use the app bundle path as its
                 # project root, i.e., to avoid 'Task assembleDebug not found'.
-                cwd=str(self.bundle_path(app)),
+                cwd=os.fsdecode(self.bundle_path(app)),
```
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/platforms/android/gradle.py:118]: Regarding `cwd`: Changed in version 3.7: cwd parameter accepts a path-like object on Windows. Shlould I remove os.fsdecode() ? 

Docs doesn't say explicitly that `subprocess.run` args can contain path-like object, but it says that: "The full function signature is largely the same as that of the Popen constructor - most of the arguments to this function are passed through to that interface." Popen args can contain path-like object. I removed `os.fsdecode` from L114, and test was successful.

File: tests/integrations/wix/test_WiX__upgrade.py [hunk 3]
```diff
@@ -82,8 +83,8 @@ def test_existing_wix_install(mock_command, tmp_path):
 
     # The download was unpacked
     mock_command.shutil.unpack_archive.assert_called_with(
-        str(wix_zip_path),
-        extract_dir=str(wix_path)
+        os.fsdecode(wix_zip_path),
+        extract_dir=os.fsdecode(wix_path)
```
[the remark this answers is not in the document]
Comment (zili55) [reply tests/integrations/wix/test_WiX__upgrade.py:87]: From docs: "Changed in version 3.7: Accepts a path-like object for filename and extract_dir." [link](https://docs.python.org/3/library/shutil.html#shutil.unpack_archive), should I remove os.fsdecode()?

File: src/briefcase/platforms/macOS/app.py [hunk 4]
```diff
@@ -45,14 +46,14 @@ def install_app_support_package(self, app: BaseConfig):
...
 
         with tempfile.TemporaryDirectory() as tmpdir:
             self.shutil.move(lib_path, tmpdir)
-            self.shutil.rmtree(str(self.support_path(app)))
+            self.shutil.rmtree(os.fsdecode(self.support_path(app)))
 
-            self.os.makedirs(str(Path(lib_path).parent))
-            self.shutil.move(str(Path(tmpdir) / 'lib'), lib_path)
+            self.os.makedirs(os.fsdecode(Path(lib_path).parent))
+            self.shutil.move(os.fsdecode(Path(tmpdir) / 'lib'), lib_path)
```
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/platforms/macOS/app.py:56]: `shutil.move` only accepts path-like objects from 3.9, I think we should use there `os.fsdecode()` [link](https://docs.python.org/3/library/shutil.html#shutil.move)

File: src/briefcase/platforms/android/gradle.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/platforms/android/gradle.py]: btw I removed all `os.fsdecode` from subprocess.run args and cwd, I will readd them if needed

File: src/briefcase/platforms/macOS/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/briefcase/platforms/macOS/app.py]: Well that's annoying :-)

In that case, you're definitely correct including the conversion here. However, it would be worth adding a docstring so future editors know (a) this isn't redundant, and (b) when we (eventually) drop 3.8 support, dropping the conversion is an option as well.

File: src/briefcase/commands/create.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/briefcase/commands/create.py:375]: As above; definitely correct to include the conversion, but we should document it for future editors.

File: src/briefcase/platforms/android/gradle.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/briefcase/platforms/android/gradle.py]: The implication of that documentation is that it won't work on *Windows* and Python 3.6, so while it will work everywhere else on 3.6, we need to retain the calls for Windows support. 

Interestingly, in my testing of `subprocess.run` on Windows with Python 3.6, `cwd` *doesn't* require the conversion, but it *is* required on the arguments; the [original 3.6 docs](https://docs.python.org/3.6/library/subprocess.html#subprocess.Popen) seem to suggest this is correct behavior for cwd, so I'm not sure what the reference in the current 3.9 docs are referring to. It works fine on Python 3.8 without the conversion in either place.

So - my inclination is to keep it for the args (or, add it back in) for the args, with a docstring indicating a cleanup when we drop 3.6.

File: tests/integrations/wix/test_WiX__upgrade.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply tests/integrations/wix/test_WiX__upgrade.py:87]: We'll need to keep this one as well; as above, add a docstring so we know we can remove it when we drop support for 3.6.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I think we're almost there. All the changes you've made look great (or you've given a good explanation why my comment was wrong). 

The handful of places where we've got "Path objects are legal, but not in Python 3.6/3.7" can be flagged with docstrings - something like:

```
# TODO: Py3.6 compatibility; os.fsencode not required in Py3.7
```

will give us something to easily search for when the time comes. If you make those handful of changes, I think we'll be good to go!

(As an aside - those docstring updates are sufficiently minor that I would normally make them myself; however, because you've made this PR off your *master* branch, I can't submit edits. As a general rule, it's a good idea to *always* do PR work in a branch - firstly, because I allows maintainers to make minor edits like these; but also because it means you can have multiple PRs in flight at the same time. Not a big deal, but if you work on another PR - and I hope you will! - it's worth keeping in mind).

File: src/briefcase/commands/create.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/commands/create.py:376]: comment added

**Comment by zili55** [comment]
@freakboy3742 You can check again the PR, if something is needed please let me know.

File: tests/integrations/xcode/test_get_simulators.py [hunk 5]
```diff
@@ -31,7 +32,7 @@ def test_simulator_is_missing(tmp_path):
     simulators = get_simulators(
         command,
         'iOS',
-        simulator_location=str(tmp_path / 'CoreSimulator.framework'),
+        simulator_location=os.fsdecode(tmp_path / 'CoreSimulator.framework'),
```
[the remark this answers is not in the document]
Comment (zili55) [reply tests/integrations/xcode/test_get_simulators.py]: `os.fsdecode` removed

File: src/briefcase/commands/create.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/commands/create.py:376]: comment added

File: src/briefcase/platforms/macOS/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zili55) [reply src/briefcase/platforms/macOS/app.py]: docstring added

**Review by freakboy3742 — APPROVED** [review]
🎉 Awesome! This looks great! Thanks for all the hard work on this - and apologies for the delay on the final review.
